### ISSUE TITLE
No limits for log record body

### ISSUE THREAD

**Issue by pellared** [open-telemetry/opentelemetry-specification#4006]
**What are you trying to achieve?**

According to the docs there are only limits for the log attributes. See: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#logrecord-limits

> LogRecord attributes MUST adhere to the [common rules of attribute limits](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute-limits).

From https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute-limits:

> If there are no limits placed on attributes, they can quickly exhaust available memory, resulting in crashes that are difficult to recover from safely.

The same problem is applicable for log record body value.

Is it intentional or an oversight? 
Should the limits be also applied for body value (even though that the whole naming is around attributes)?
Should we have different limits for log record value?

**Comment by pellared** [comment]
PTAL @open-telemetry/specs-logs-approvers

**Comment by jack-berg** [comment]
We might not want the limits by default, but I think its reasonable to have them available. Currently, its not possible to configure any limits for the body, and its cumbersome to do in the collector ([see this slack convo](https://cloud-native.slack.com/archives/C024DRQ63UP/p1687896379414009?thread_ts=1686585801.106229&cid=C024DRQ63UP) - it previously wasn't possible). 

> Related, how would these limits apply to a non-structured body?

The key question is how do attribute limits work on an AnyValue type. Note this will be relevant for events as well.

The [common attribute limits](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common#attribute-limits) provide configuration for the max allowed attribute count, and the maximum attribute length limit.

[AnyValue](https://github.com/open-telemetry/opentelemetry-proto/blob/342e1d4c3a1fe43312823ffb53bd38327f263059/opentelemetry/proto/common/v1/common.proto#L28-L67) is a recursive data structure, and the common attribute limits were designed for a flat list of key value pairs. Here's how we might apply the rules to AnyValue:

- AnyValue is oneof: string_value, bool_value, int_value, double_value, array_value, kv_list_value, bytes_value
  - If bool_value, int_value, double_value, ignore attribute limits
  - If string_value, bytes_value, truncate the value to the `AttributeValueLengthLimit`
  - If kv_list_value, limit the entires to `AttributeCountLimit`, apply limits recursively to each AnyValue value
  - If array_value, apply limits recursively to each AnyValue value

Note that the current rules don't limit the number of entries in an array, which is odd. I've carried that rule forward into the proposal above, but perhaps there should be a third property which limits the number of array entries.

**Comment by pellared** [comment]
@jack-berg, I do not think that it would be good to use `AttributeValueLengthLimit` and `AttributeCountLimit` for `Body`.  Reasons:

1. "we probably don't need to apply the same strict limits, though other limits may be necessary." by @djaglowski 
2. While Body uses the same "structure" as Attributes they are not semantically equivalent therefore I do not think we should use a single configuration to control both of them.

For now, (as a baby-step) I simply propose to clarify in the specification that the attribute limits do not apply on the Body.

**Comment by svrnm** [comment]
@pellared @jack-berg please take a look, is this still needed? how can we move this forward?

**Comment by pellared** [comment]
Hard for me to evaluate as Logs are still Beta in OTel Go and I personally, I have not feedback on how much this feature is needed. I think we can still leave a "community-feedback".

**Comment by pellared** [comment]
> [AnyValue](https://github.com/open-telemetry/opentelemetry-proto/blob/342e1d4c3a1fe43312823ffb53bd38327f263059/opentelemetry/proto/common/v1/common.proto#L28-L67) is a recursive data structure, and the common attribute limits were designed for a flat list of key value pairs. Here's how we might apply the rules to AnyValue:
> 
>     * AnyValue is oneof: string_value, bool_value, int_value, double_value, array_value, kv_list_value, bytes_value
>       
>       * If bool_value, int_value, double_value, ignore attribute limits
>       * If string_value, bytes_value, truncate the value to the `AttributeValueLengthLimit`
>       * If kv_list_value, limit the entires to `AttributeCountLimit`, apply limits recursively to each AnyValue value
>       * If array_value, apply limits recursively to each AnyValue value


This was done in https://github.com/open-telemetry/opentelemetry-specification/pull/4651 without 

> limit the entires to `AttributeCountLimit` 

because there was no consensus so far on 

> Note that the current rules don't limit the number of entries in an array, which is odd. I've carried that rule forward into the proposal above, but perhaps there should be a third property which limits the number of array entries.

Currently the new limit rules are in "Development", the plan is to stabilize them as part of https://github.com/open-telemetry/opentelemetry-specification/issues/4710.

I updated the issue description as we still have no limits for `Body` (as limits apply only within attributes), `SeverityText`, and map/array count of `AnyValue`.

**Comment by philsttr** [comment]
I'd like to provide a related real-world example for consideration.

Using the OpenTelemetry SDK Java 1.58.0, I have configured:
* `otel.attribute.value.length.limit=4096`
* `otel.attribute.count.limit=128` (using default value)

I recently ran into the following error when sending logs to Grafana Loki's OTLP endpoint.

```
Failed to export logs. Server responded with HTTP status code 400. Error message: 1 errors like: stream '{...redacted...}' has structured metadata too large: '2974287' bytes, limit: '65536' bytes. Please see `limits_config.max_structured_metadata_size` or contact your Loki administrator to increase it; 1 errors like: stream 
```

I intended the otel limits that I put in place to prevent the structured metadata (aka attributes on otel side) from being too large.

After some debugging, I found some code that was adding a large List as an attribute value.

Since otel enforces the attribute value length  _for each entry_ rather than _the list as a whole_, otel did not truncate the attribute value to 4k as desired, which lead to the huge attribute value, which led to the rejection from Loki.


Ideally, I would like some way to guarantee that events sent to a backend can be limited/truncated to sizes less than my backend's limits  (therefore, guaranteeing that events will never be rejected by my backend due to size restrictions).   That means placing various limits on the sending side, so things will be truncated.

In the example above, I would like a true "attribute value length limit" to apply to the full "attribute value" (i.e. the full list),  not the "attribute sub-values" (i.e. each element in a list).  This would mean either changing the existing behavior of `attribute.value.length.limit` for list attribute values or introducing another limit that applies to the whole list.

In addition, I would like to limit the body as well (not shown in example, but could also cause backends to reject an event)

**Comment by pellared** [comment]
Discussed during the 2026-02-17 Logs SIG meeting.

We most likely do not need an `SeverityText` limit similarly we do not have a limit for span name. If we need a limit for this then we would need to have a limit for everything. 

We may need to have some limits that accumulates all nodes of complex attributes (maybe even a limit for the whole log record - including Body, SeverityText). This would need to be new limit properties as reusing the existing once can break existing users. We cannot change its existing semantics which are derived from "if it is an array of strings, then apply the limit to each value within the array separately" that were before "complex attributes" were introduced.

**Comment by pellared** [comment]
Regarding " map/array count of AnyValue" and

> We may need to have some limits that accumulates all nodes of complex attribute

This is one way of addressing this https://github.com/open-telemetry/opentelemetry-specification/pull/5186
